src/oauth.rs

```diff
@@ -0,0 +1,107 @@
+use crate::token::TemporaryToken;
+use reqwest::{Client, Method};
+use snafu::{ResultExt, Snafu};
+use std::time::{Duration, Instant};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to decode service account key: {}", source))]
+    DecodeKey { source: jsonwebtoken::errors::Error },
+
+    #[snafu(display("Unable to encode jwt: {}", source))]
+    EncodeJwt { source: jsonwebtoken::errors::Error },
+
+    #[snafu(display("Error performing token request: {}", source))]
+    TokenRequest { source: reqwest::Error },
+}
+
+#[derive(serde::Serialize)]
+struct TokenClaims<'a> {
+    iss: &'a str,
+    scope: &'a str,
+    aud: &'a str,
+    exp: u64,
+    iat: u64,
+}
+
+#[derive(serde::Deserialize, Debug)]
+struct TokenResponse {
+    access_token: String,
+    expires_in: u64,
+}
+
+/// Encapsulates the logic to perform an OAuth token challenge
+#[derive(Debug)]
+pub struct OAuthProvider {
```

Comment:
This will also be useful for Azure which similarly adopts standard authentication practices. AWS unfortunately is a special snowflake here, as per usual

Reply:
:D